napi-inl.h

```diff
@@ -3697,10 +3723,31 @@ inline napi_value ObjectWrap<T>::ConstructorCallbackWrapper(
   }
 
   T* instance;
-  napi_value wrapper = details::WrapCallback([&] {
+  napi_value wrapper = details::WrapCallback([&] () -> napi_value {
     CallbackInfo callbackInfo(env, info);
+    FinalizerHint::Init(callbackInfo);
+#ifdef NAPI_CPP_EXCEPTIONS
+    try {
+      instance = new T(callbackInfo);
+      return callbackInfo.This();
+    }
+    catch (...) {
+      FinalizerHint* finalizerHint = FinalizerHint::Get(callbackInfo);
+      if (finalizerHint != nullptr) {
+         finalizerHint->shouldFinalize = false;
+      }
+      throw;
+    }
+#else
     instance = new T(callbackInfo);
+    if (callbackInfo.Env().IsExceptionPending()) {
+      FinalizerHint* finalizerHint = FinalizerHint::Get(callbackInfo);
+      if (finalizerHint != nullptr) {
```

Comment:
This check should be a hard check. If `finalizerHint` is NULL, we must go to `Napi::Error::Fatal`. After all, we set it above.